testing/unittests/jlibtests.cpp

```diff
@@ -3396,7 +3396,308 @@ element1: !el scalarvalue
+
+            // ValueTypes
+            addValueType(compat, "secret", "r");
+            addValueType(compat, "secret-if-a", "r", true, "value-type-set-a");
+            addValueType(compat, "secret-if-b", "r", false, "value-type-set-b");
+            addValueType(compat, "*", "r");
+
+            // Rules
+            addRule(compat, "", "r");
+            addRule(compat, "xml", "r");
+        }
+
+        // Second compatibility element
+        {
+            IPropertyTree *compatibility = root->addPropTree("__item__");
+            IPropertyTree *compat = compatibility->addPropTree("compatibility");
+
+            IPropertyTree *context = compat->addPropTree("context");
+            context->setProp("@domain", "urn:hpcc:unittest");
+            context->setProp("@version", "0");
+            addProperty(context, "valuetype-set", "value-type-set-a");
+            addProperty(context, "rule-set", "");
+
+            addValueType(compat, "secret", "r");
+            addValueType(compat, "secret-if-a", "r", true, "value-type-set-a");
+            addValueType(compat, "secret-if-b", "p", false, "value-type-set-b");
+
+            addRule(compat, "", "r");
+            addRule(compat, "xml", "r");
+        }
+
+        // Third compatibility element
+        {
+            IPropertyTree *compatibility = root->addPropTree("__item__");
+            IPropertyTree *compat = compatibility->addPropTree("compatibility");
 
+            IPropertyTree *context = compat->addPropTree("context");
+            context->setProp("@domain", "urn:hpcc:unittest");
+            context->setProp("@version", "0");
+            addProperty(context, "valuetype-set"); // No value attribute
+            addProperty(context, "rule-set", "rule-set-2");
+
+            addValueType(compat, "secret", "r");
+            addValueType(compat, "secret-if-a", "p", true, "value-type-set-a");
+            addValueType(compat, "secret-if-b", "p", false, "value-type-set-b");
+
+            addRule(compat, "", "p");
+            addRule(compat, "xml", "p");
+        }
+
+        // Fourth compatibility element
+        {
+            IPropertyTree *compatibility = root->addPropTree("__item__");
+            IPropertyTree *compat = compatibility->addPropTree("compatibility");
+
+            IPropertyTree *context = compat->addPropTree("context");
+            context->setProp("@domain", "urn:hpcc:unittest");
+            context->setProp("@version", "0");
+            addProperty(context, "valuetype-set", "value-type-set-b");
+            addProperty(context, "rule-set", "rule-set-2");
+
+            addValueType(compat, "secret", "r");
+            addValueType(compat, "secret-if-a", "p", true, "value-type-set-a");
+            addValueType(compat, "secret-if-b", "r", false, "value-type-set-b");
+
+            addRule(compat, "", "r");
+            addRule(compat, "xml", "r");
+        }
+
+        return LINK(root.get());
+    }
+
+    IPropertyTree *createCPTValuePTree()
```

Comment:
all trees have CPTValue's, so this is a bit misleading.
I would be better to emphasis that they're binary blobs , and some large and compressed